api/v1alpha1/amaltheasession_children.go

```diff
@@ -0,0 +1,220 @@
+	labels := labelsForAmaltheaSession(cr.Name)
+	replicas := int32(1)
+	if cr.Spec.Hibernated {
+		replicas = 0
+	}
+
+	session := cr.Spec.Session
+	pvc := cr.PVC()
+
+	sessionContainer := v1.Container{
+		Image:           session.Image,
+		Name:            "session",
+		ImagePullPolicy: v1.PullIfNotPresent,
+
+		Ports: []v1.ContainerPort{{
+			ContainerPort: session.Port,
+			Name:          "session-port",
+			Protocol:      v1.ProtocolTCP,
+		}},
+
+		Args:                     session.Args,
+		Command:                  session.Command,
+		Env:                      session.Env,
+		Resources:                session.Resources,
+		VolumeMounts:             []v1.VolumeMount{{Name: "session", MountPath: session.Storage.MountPath}},
+		TerminationMessagePath:   "/dev/termination-log",
+		TerminationMessagePolicy: v1.TerminationMessageReadFile,
+	}
+
+	securityContext := &v1.SecurityContext{
+		RunAsNonRoot: &[]bool{true}[0],
+		RunAsUser:    &[]int64{session.RunAsUser}[0],
+		RunAsGroup:   &[]int64{session.RunAsGroup}[0],
+	}
+
+	if session.RunAsUser == 0 {
+		securityContext.RunAsNonRoot = &[]bool{false}[0]
+	}
+
+	sessionContainer.SecurityContext = securityContext
+
+	containers := []v1.Container{sessionContainer}
+	containers = append(containers, cr.Spec.ExtraContainers...)
+
+	sts := appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            cr.Name,
+			Namespace:       cr.Namespace,
+			OwnerReferences: []metav1.OwnerReference{cr.OwnerReference()},
+		},
+		Spec: appsv1.StatefulSetSpec{
+			Replicas: &replicas,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: labels,
+			},
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: labels,
+				},
+				Spec: v1.PodSpec{
+					Containers:     containers,
+					InitContainers: cr.Spec.ExtraInitContainers,
+					Volumes:        []v1.Volume{{Name: "session", VolumeSource: v1.VolumeSource{PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{ClaimName: pvc.Name}}}},
+				},
+			},
+		},
+	}
+	return sts
+}
+
+// Service returns a AmaltheaSession Service object
+func (cr *AmaltheaSession) Service() v1.Service {
+	labels := labelsForAmaltheaSession(cr.Name)
+
+	svc := v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            cr.Name,
+			Namespace:       cr.Namespace,
+			OwnerReferences: []metav1.OwnerReference{cr.OwnerReference()},
+		},
+		Spec: v1.ServiceSpec{
+			Selector: labels,
+			Ports: []v1.ServicePort{{
+				Protocol:   v1.ProtocolTCP,
+				Name:       "session-port",
+				Port:       80,
+				TargetPort: intstr.FromInt32(cr.Spec.Session.Port),
+			}},
+		},
+	}
+	return svc
+}
+
+// Ingress returns a AmaltheaSession Ingress object
+func (cr *AmaltheaSession) Ingress() *networkingv1.Ingress {
+	labels := labelsForAmaltheaSession(cr.Name)
+
+	ingress := cr.Spec.Ingress
+	if reflect.DeepEqual(ingress, Ingress{}) {
+		return nil
+	}
+
+	path := "/"
+	if ingress.PathPrefix != "" {
+		path = ingress.PathPrefix
+	}
```

Comment:
I am just realizing, shouldn't we rather have PathPrefix defined with a default at "/" ?

Reply:
done